src/adaptors/mod.rs

```diff
@@ -753,53 +753,88 @@ impl<I, F> Iterator for Coalesce<I, F>
     }
 }
 
-/// An iterator adaptor that removes repeated duplicates.
+/// An iterator adaptor that removes repeated duplicates, determining equality using a comparison function.
 ///
-/// See [`.dedup()`](../trait.Itertools.html#method.dedup) for more information.
+/// See [`.dedup_by()`](../trait.Itertools.html#method.dedup_by) or [`.dedup()`](../trait.Itertools.html#method.dedup) for more information.
 #[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
-pub struct Dedup<I>
+pub struct DedupBy<I, Cmp>
     where I: Iterator
 {
     iter: CoalesceCore<I>,
+    cmp: Cmp,
+}
+
+pub trait CompareEq<T> { // TODO replace by Fn(&T, &T)->bool once Rust supports it
+    fn cmp_eq(&mut self, a: &T, b: &T) -> bool;
 }
 
-impl<I: Clone> Clone for Dedup<I>
+#[derive(Clone)]
+pub struct SimpleCompareEq;
```

Comment:
This is marked `pub`, but it isn't actually re-exported from adaptors in `lib.rs`:
https://github.com/rust-itertools/itertools/blob/3296797f4faf525c24282faefa12ea6f45ac639f/src/lib.rs#L80-L97

Is this intentional? Does it pose a usability hazard? At the very least, it means `SimpleCompareEq` isn't documented (and thus isn't clickable in the documentation for `Dedup`).

(The same concern applies to the `CompareEq` trait.)

I'm not necessarily opposed to using the private-in-public trick to keep the public API surface small, but I want to make sure we don't do so unintentionally.

Reply:
I am not sure about this, but I think I need them to be `pub` because they are exported from `adaptors` and private types seem to be disallowed in public structs.

Reply:
Yes, they need to be `pub`.

I think I'm satisfied not re-exporting them for the time being.